### PR TITLE
Add dockerfile to create custom spark image for CDAP execution

### PR DESCRIPTION
**Pull request by albertshau**
Added build steps to create a docker image that contains a
customized spark image that can be used to run CDAP spark programs

### CODE REVIEW

File: cdap-spark-core-base/src/k8s/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# docker build --no-cache --build-arg base_image=<image> \
+#   --build-arg cdap_version=`mvn -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec` \
+#   -f cdap-spark-core-base/src/k8s/Dockerfile \
+#   . -t <image>:<tag>
+
+ARG base_image
+
+FROM ${base_image}
+
+ARG cdap_spark_module=cdap-spark-core3_2.12
```
Comment (chtyim) [inline cdap-spark-core-base/src/k8s/Dockerfile]: Should this not be hardcoded?

File: cdap-spark-core-base/src/k8s/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ARG cdap_version
+
+# copy cdap-spark-core and its dependencies
+COPY ${cdap_spark_module}/target/k8s/${cdap_spark_module} /opt/cdap/cdap-spark-core
```
Comment (chtyim) [inline cdap-spark-core-base/src/k8s/Dockerfile]: I would suggest have the maven to output to `${cdap_spark_module}/target/k8s`. We shouldn't need the `${cdap_spark_module}` suffix again

File: cdap-spark-core-base/src/k8s/Dockerfile (hunk 1 above)
Comment (albertshau) [reply cdap-spark-core-base/src/k8s/Dockerfile]: yeah, was originally thinking we could have a similar spark2 one, but that didn't turn out. will remove

File: cdap-spark-core-base/src/k8s/Dockerfile (hunk 2 above)
Comment (albertshau) [reply cdap-spark-core-base/src/k8s/Dockerfile]: 👍

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java [hunk 3]
```diff
@@ -63,22 +63,44 @@ public boolean acceptPackage(String packageName) {
...
+  }
+
+  /**
+   * Launches the given main class. The main class will be loaded through the {@link SparkContainerClassLoader}.
+   *
+   * @param mainClassName the main class to launch
+   * @param args arguments for the main class
+   * @param removeMainClass whether to remove the jar for the main class from the classloader
+   */
+  @SuppressWarnings("unused")
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: This method is used by the `main` and the other `launch` method, right?

**Review by chtyim — APPROVED** [review]

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java (hunk 3 above)
Comment (albertshau) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: ah yes, I will remove the annotation

**Pull request merged**